haystack/document_stores/opensearch.py

```diff
@@ -798,6 +1005,126 @@ def _get_raw_similarity_score(self, score):
+                "training_field": self.embedding_field,
+                "dimension": self.embedding_dim,
+                "method": {
+                    "name": "ivf",
+                    "engine": "faiss",
+                    "space_type": self.space_type,
+                    "parameters": {"nlist": nlist, "nprobes": nprobes},
+                },
+            }
+            # Add product quantization
+            if self.index_type == "ivf_pq":
+                m = 1 if "m" not in self.knn_parameters else self.knn_parameters["m"]
+                code_size = 8 if "code_size" not in self.knn_parameters else self.knn_parameters["code_size"]
+                if code_size > 8:
+                    raise DocumentStoreError(
+                        f"code_size parameter for product quantization must be less than or equal to 8. "
+                        f"Provided code_size is `{code_size}`."
+                    )
+
+                # see FAISS doc for details: https://github.com/facebookresearch/faiss/wiki/FAQ#can-i-ignore-warning-clustering-xxx-points-to-yyy-centroids
+                min_points_per_cluster = 39
+                n_clusters = 2**code_size
+                if len(documents) < n_clusters:
+                    raise DocumentStoreError(
+                        f"PQ training requires the number of training samples to be greater than or "
+                        f"equal to the number of clusters. Number of provided training samples is `{len(documents)}` "
+                        f"and the number of clusters is `{n_clusters}`."
+                    )
+                if len(documents) < n_clusters * min_points_per_cluster:
+                    logger.warning(
+                        "Consider increasing the number of training samples to at least "
+                        "`%i` to get a reliable PQ index.",
+                        n_clusters * min_points_per_cluster,
+                    )
+
+                encoder = {"name": "pq", "parameters": {"m": m, "code_size": code_size}}
+                settings += f" m:{m} code_size:{code_size}"
+                training_req_body["method"]["parameters"]["encoder"] = encoder
+
+            training_req_body["description"] = settings
+            logger.info("Training IVF index '%s' using {len(documents)} embeddings.", index)
+            train_endpoint = f"/_plugins/_knn/models/{index}-ivf/_train"
+            response = self.client.transport.perform_request(
+                "POST", url=train_endpoint, headers=headers, body=training_req_body
+            )
+            ivf_model = response["model_id"]
+
+            # Wait until model training is finished
+            model_state_endpoint = f"/_plugins/_knn/models/{ivf_model}"
+            response = self.client.transport.perform_request("GET", url=model_state_endpoint, headers=headers)
+            model_state = response["state"]
+            while model_state != "created":
```

Comment:
Can we use some delay here? This could flood the OpenSearch instance a bit. E.g. tenacy.retry should solve this easily.